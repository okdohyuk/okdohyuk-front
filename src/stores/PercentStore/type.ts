export interface PercentCalculator {
  primaryNumber: number | string;
  secondaryNumber: number | string;
  result: number | string;
}

export interface PercentageUpDown extends PercentCalculator {
  isIncrease: boolean;
}

export interface PercentCalculators {
  percentageOfTotal: PercentCalculator;
  partOfTotal: PercentCalculator;
  findPercentage: PercentCalculator;
  percentageUpDown: PercentageUpDown;
  findPercentageValue: PercentCalculator;
}

export interface ValueChange {
  target: keyof PercentCalculators;
  targetValue: keyof PercentCalculator | keyof PercentageUpDown;
  value: string;
}

export interface PercentStoreState {
  calculators: PercentCalculators;
  valueChange: (props: ValueChange) => void;
  handleResultCalculator: (target: keyof PercentCalculators) => void;
}
