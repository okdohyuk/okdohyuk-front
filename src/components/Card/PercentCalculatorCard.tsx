import React from 'react';
import {
  PercentCalculator,
  PercentageUpDown,
  PercentStoreState,
  ValueChange,
} from '@stores/PercentStore';

type PercentCalculatorCard = {
  calculator: PercentCalculator | PercentageUpDown;
  calculatorName: keyof PercentStoreState;
  valueChange: (props: ValueChange) => void;
  title: string;
  placeholder: string[];
  text: string[];
};

function PercentCalculatorCard({
  calculatorName,
  calculator,
  title,
  valueChange,
  placeholder,
  text,
}: PercentCalculatorCard) {
  return (
    <div className={'w-full flex flex-col space-y-4 rounded-xl bg-zinc-300 dark:bg-zinc-800 p-5'}>
      <div className={'flex'}>
        <h3 className={'dark:text-white'}>{title}</h3>
      </div>
      <div className={'w-full flex justify-between'}>
        <div className={'w-2/3 flex items-center space-x-1'}>
          <input
            type={'text'}
            className={'w-1/3 h-[50px] rounded-2xl p-2 text-right'}
            placeholder={placeholder[0]}
            value={calculator.primaryNumber}
            onChange={(e) =>
              valueChange({
                target: calculatorName,
                targetValue: 'primaryNumber',
                value: e.target.value,
              })
            }
          />
          <span className={'dark:text-white'}>{text[0]}</span>
          <input
            type={'text'}
            className={'w-1/3 h-[50px] rounded-2xl p-2 text-right'}
            placeholder={placeholder[1]}
            value={calculator.secondaryNumber}
            onChange={(e) =>
              valueChange({
                target: calculatorName,
                targetValue: 'secondaryNumber',
                value: e.target.value,
              })
            }
          />
          <span className={'dark:text-white'}>{text[1]}</span>
          {'isIncrease' in calculator ? (
            <>
              <select
                className={'w-1/3 h-[50px] rounded-2xl p-2 text-right'}
                value={calculator.isIncrease ? 'true' : 'false'}
                onChange={(e) =>
                  valueChange({
                    target: calculatorName,
                    targetValue: 'isIncrease',
                    value: e.target.value,
                  })
                }
              >
                <option value={'true'}>증가</option>
                <option value={'false'}>감소</option>
              </select>
              <span className={'dark:text-white'}>{'하면?'}</span>
            </>
          ) : null}
        </div>
        <div className={'w-1/4 flex items-center space-x-1'}>
          <span className={'dark:text-white'}>{text[2]}</span>
          <input
            type={'text'}
            className={'w-full h-[50px] rounded-2xl p-2 text-right'}
            placeholder={placeholder[2]}
            value={calculator.result.toLocaleString()}
            onChange={() => calculator.result}
          />
          <span className={'dark:text-white'}>{text[3]}</span>
        </div>
      </div>
    </div>
  );
}

export default PercentCalculatorCard;
