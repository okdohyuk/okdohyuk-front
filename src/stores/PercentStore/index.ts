import { observable, action, makeObservable } from 'mobx';
import React from 'react';

interface PercentCalculator {
  primaryNumber: number | string;
  secondaryNumber: number | string;
  result: number | string;
}

interface PercentageUpDown extends PercentCalculator {
  isIncrease: boolean;
}

interface PercentStoreState {
  percentageOfTotal: PercentCalculator;
  partOfTotal: PercentCalculator;
  findPercentage: PercentCalculator;
  percentageUpDown: PercentageUpDown;
  findPercentageValue: PercentCalculator;
  valueChange: (props: ValueChange) => void;
}

type ValueChange = {
  target: keyof PercentStoreState;
  targetValue: keyof PercentCalculator;
  value: React.ChangeEvent<HTMLInputElement>;
};

const initialValue: PercentCalculator = {
  primaryNumber: '',
  secondaryNumber: '',
  result: '',
};

class PercentStore implements PercentStoreState {
  @observable public percentageOfTotal: PercentCalculator = initialValue;
  @observable public partOfTotal: PercentCalculator = initialValue;
  @observable public findPercentage: PercentCalculator = initialValue;
  @observable public percentageUpDown: PercentageUpDown = { ...initialValue, isIncrease: true };
  @observable public findPercentageValue: PercentCalculator = initialValue;

  constructor() {
    makeObservable(this);
  }

  @action public valueChange = (props: ValueChange) => {
    if (props.target === 'valueChange') return;
    if (isNaN(+props.value.target.value)) return;
    this[props.target][props.targetValue] = +props.value.target.value;

    if (isNaN(+this[props.target].primaryNumber) || isNaN(+this[props.target].secondaryNumber))
      return;

    this[props.target].result =
      (+this[props.target].primaryNumber * +this[props.target].secondaryNumber) / 100;
  };
}

export default PercentStore;
