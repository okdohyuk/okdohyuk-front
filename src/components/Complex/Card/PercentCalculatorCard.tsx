import React, { useMemo } from 'react';
import { PercentCalculators } from '@stores/PercentStore/type';
import { toJS } from 'mobx';
import useStore from '@hooks/useStore';
import { observer } from 'mobx-react';
import { useTranslation } from 'next-i18next';
import PercentStore from '@stores/PercentStore';

type PercentCalculatorCard = {
  calculatorName: keyof PercentCalculators;
  title: string;
  placeholder: string[];
  text: string[];
};

function PercentCalculatorCard({
  calculatorName,
  title,
  placeholder,
  text,
}: PercentCalculatorCard) {
  const percentStore = useStore<PercentStore>('percentStore');
  const { valueChange } = percentStore;
  const { calculators } = toJS(percentStore);
  const { t } = useTranslation('percent');

  const calculator = useMemo(() => {
    return calculators[calculatorName];
  }, [calculatorName, calculators]);

  return (
    <div className={'w-full flex flex-col space-y-4 rounded-md bg-basic-4 p-4'}>
      <div className={'flex'}>
        <h3 className={'t-d-1 t-basic-1'}>{title}</h3>
      </div>
      <div className={'w-full flex flex-col md:flex-row justify-between space-y-1 md:space-y-0'}>
        <div className={'w-full md:w-3/4 flex items-center space-x-1'}>
          <input
            type={'text'}
            inputMode={'numeric'}
            className={'input-text w-1/4 text-right'}
            placeholder={placeholder[0]}
            value={calculator.primaryNumber}
            onChange={(e) =>
              valueChange({
                target: calculatorName,
                targetValue: 'primaryNumber',
                value: e.target.value,
              })
            }
          />
          <span className={'t-d-2 t-basic-1'}>{text[0]}</span>
          <input
            type={'text'}
            inputMode={'numeric'}
            className={'input-text w-1/4 text-right'}
            placeholder={placeholder[1]}
            value={calculator.secondaryNumber}
            onChange={(e) =>
              valueChange({
                target: calculatorName,
                targetValue: 'secondaryNumber',
                value: e.target.value,
              })
            }
          />
          <span className={'t-d-2 t-basic-1'}>{text[1]}</span>
          {'isIncrease' in calculator ? (
            <>
              <select
                className={'input-text w-1/6 text-right'}
                value={calculator.isIncrease ? 'true' : 'false'}
                onChange={(e) =>
                  valueChange({
                    target: calculatorName,
                    targetValue: 'isIncrease',
                    value: e.target.value,
                  })
                }
              >
                <option value={'true'}>{t('increase')}</option>
                <option value={'false'}>{t('decrease')}</option>
              </select>
              <span className={'t-d-2 t-basic-1'}>{t('to')}</span>
            </>
          ) : null}
        </div>
        <div className={'w-1/3 md:w-1/4 flex ml-auto md:ml-0 items-center space-x-1'}>
          <span className={'t-d-2 t-basic-1'}>{text[2]}</span>
          <input
            type={'text'}
            inputMode={'none'}
            className={'input-text w-full text-right'}
            placeholder={placeholder[2]}
            value={calculator.result.toLocaleString()}
            onChange={() => calculator.result}
          />
          <span className={'t-d-2 t-basic-1'}>{text[3]}</span>
        </div>
      </div>
    </div>
  );
}

export default observer(PercentCalculatorCard);
