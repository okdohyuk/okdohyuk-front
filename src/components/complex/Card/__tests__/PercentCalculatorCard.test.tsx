import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import PercentCalculatorCard from '@components/legacy/complex/Card/PercentCalculatorCard';
import PercentStore from '@stores/PercentStore';
import { Provider } from 'mobx-react';

type TestElement = Document | Element | Window | Node;

function hasInputValue(e: TestElement, inputValue: string) {
  return screen.getByDisplayValue(inputValue) === e;
}

describe('<PercentCalculatorCard />', () => {
  it('matches snapshot', () => {
    const percentStore = new PercentStore();
    const utils = render(
      <Provider percentStore={percentStore}>
        <PercentCalculatorCard
          title={'전체값의 몇 퍼센트는 얼마인가 계산'}
          placeholder={['10000', '20', '2,000']}
          calculatorName={'percentageOfTotal'}
          text={['의', '%', '=']}
        />
      </Provider>,
    );
    expect(utils.container).toMatchSnapshot();
  });
  it('calculation', () => {
    const percentStore = new PercentStore();
    render(
      <Provider percentStore={percentStore}>
        <PercentCalculatorCard
          title={'전체값의 몇 퍼센트는 얼마인가 계산'}
          placeholder={['10000', '20', '2,000']}
          calculatorName={'percentageOfTotal'}
          text={['의', '%', '=']}
        />
      </Provider>,
    );
    const inputOne = screen.getByPlaceholderText('10000');
    const inputTwo = screen.getByPlaceholderText('20');
    const result = screen.getByPlaceholderText('2,000');
    fireEvent.change(inputOne, { target: { value: '10000' } });
    fireEvent.change(inputTwo, { target: { value: '20' } });
    expect(hasInputValue(inputOne, '10000')).toBe(true);
    expect(hasInputValue(inputTwo, '20')).toBe(true);
    expect(hasInputValue(result, '2,000')).toBe(true);
  });
});
