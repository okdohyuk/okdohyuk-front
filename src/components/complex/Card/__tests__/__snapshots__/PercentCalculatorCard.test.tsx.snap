// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PercentCalculatorCard /> matches snapshot 1`] = `
<div>
  <div
    class="w-full flex flex-col space-y-4 rounded-md bg-basic-4 p-4"
  >
    <div
      class="flex"
    >
      <h3
        class="t-d-1 t-basic-1"
      >
        전체값의 몇 퍼센트는 얼마인가 계산
      </h3>
    </div>
    <div
      class="w-full flex flex-col md:flex-row justify-between space-y-1 md:space-y-0"
    >
      <div
        class="w-full md:w-3/4 flex items-center space-x-1"
      >
        <input
          class="input-text w-1/4 text-right"
          inputmode="numeric"
          placeholder="10000"
          type="text"
          value=""
        />
        <span
          class="t-d-2 t-basic-1"
        >
          의
        </span>
        <input
          class="input-text w-1/4 text-right"
          inputmode="numeric"
          placeholder="20"
          type="text"
          value=""
        />
        <span
          class="t-d-2 t-basic-1"
        >
          %
        </span>
      </div>
      <div
        class="w-1/3 md:w-1/4 flex ml-auto md:ml-0 items-center space-x-1"
      >
        <span
          class="t-d-2 t-basic-1"
        >
          =
        </span>
        <input
          class="input-text w-full text-right"
          inputmode="none"
          placeholder="2,000"
          type="text"
          value=""
        />
        <span
          class="t-d-2 t-basic-1"
        />
      </div>
    </div>
  </div>
</div>
`;
